src/back/spv/writer.rs

```diff
@@ -1612,23 +1641,82 @@ impl Writer {
 
     fn write_block(
         &mut self,
-        statements: &[crate::Statement],
+        statements: &mut Vec<crate::Statement>,
         ir_module: &crate::Module,
         ir_function: &crate::Function,
         function: &mut Function,
-    ) -> Result<spirv::Word, Error> {
+        branch_id: Option<spirv::Word>,
+    ) -> Result<Block, Error> {
         let mut block = Block::default();
-        let id = self.generate_id();
-        block.label = Some(super::instructions::instruction_label(id));
 
-        for statement in statements {
-            match *statement {
+        while !statements.is_empty() {
```

Comment:
currently, I'd like to understand why this needs to be a vector we are mutating, before getting to the rest

Reply:
Yes, I understand and I am not entirely sure if this is the right way code wise and memory wise. What I want to do is iterating through the statements while it is not empty. 

When does it become empty? By either organically iterating over it and removing the first item (like a dequeue, see line 1655), or via a branching.

When we reach branching, we want to essentially "move" the remaining statements to the post dominator/merge block (block after the branching), and continue to process the remaining statements.

As this is a recursive function, the first function call will be the first block and when the branching and merging is finished, it doesn't have any statements left (as it is drained via the `create_post_dominate_block()` function.

Like I said, I am not sure if this is the right way, hence I think this is not a good way, but this was the first solution that came into my mind, without considering memory footprints etc.

Reply:
I'm not sure I understand why anything needs to be moved. Let's say you are processing a statement somewhere. In the context for this, you need to know the label ID you are in, and the label ID where you'd need to merge to. If you encounter an if/else statement, you'd create those new block labels, call yourself recursively for true/false statements (in which case you know exactly what the merge point is, and what the parent ID is). Then you'd just continue as usual through the rest of the statements, just with the "my block ID" pointing to the new label you created for the stuff after if/else. So it's supposed to be just a routine that recursively processes the statements and spawns the necessary blocks. The only little unspecified bit here is who is responsible for creation of which block labels, but the answer to that will come naturally as you are implementing the code.

Am I terribly missing something here?